File: first_pass/easy/Valid Parentheses/isValid.js
/**
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', 
 * determine if the input string is valid.
 * 
 * An input string is valid if:
 * 
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Every close bracket has a corresponding open bracket of the same type.
 */

/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function(s) {
    const stack = [];
    const n = s.length;

    for (let i = 0; i < n; ++i) {
        const parenthesis = s[i];

        if (parenthesis == ")" || parenthesis == "]" || parenthesis == "}") {
            const other = stack.pop();
            switch (parenthesis) {
                case ")": 
                    if (other != "(") return false;
                    break;
                case "}": 
                    if (other != "{") return false;
                    break;
                case "]": 
                    if (other != "[") return false;
                    break;
                default:
                    return false;
            }
        } else {
            stack.push(parenthesis);
        }
    }

    return stack.length === 0
};

const testIsValid = (s, output) => {
    const result = isValid(s);
    if (result === output) {
        console.log("Test Case Passed!");
    } else {
        console.log(
            `Test Case Failed: s: ${s} output: ${output} result: ${result}`
        );
    }
}

testIsValid("()", true);
testIsValid("()[]{}", true);
testIsValid("(]", false)

File: second_pass/4_stack/1_valid_parentheses/is_valid.py
class MySolution:
    def isValid(self, s: str) -> bool:
        stack = []
        n = len(s)

        for i in range(n):
            if s[i] in ["(", "{", "["]:
                stack.append(s[i])
            elif stack:
                prev = stack.pop()

                if s[i] == ")" and prev != "(":
                    return False
                if s[i] == "}" and prev != "{":
                    return False
                if s[i] == "]" and prev != "[":
                    return False
            else:
                return False

        return len(stack) == 0

    def testIsValid(self, s, expected):
        actual = self.isValid(s)
        if actual == expected:
            print("Test Case Passed!")
        else:
            print(f"Test Case Failed! s: {s}, actual: {actual}, expected: {expected}")


MySolution().testIsValid("()", True)
MySolution().testIsValid("()[]{}", True)
MySolution().testIsValid("(]", False)
MySolution().testIsValid("]", False)
